libtenzir/builtins/formats/feather2.cpp

```diff
@@ -0,0 +1,203 @@
+
+class CallbackListener : public arrow::ipc::Listener {
+public:
+  CallbackListener() = default;
+
+  arrow::Status OnRecordBatchDecoded(
+    std::shared_ptr<arrow::RecordBatch> record_batch) override {
+    record_batch_buffer.push(record_batch);
+    return arrow::Status::OK();
+  }
+
+  std::queue<std::shared_ptr<arrow::RecordBatch>> record_batch_buffer;
+};
+
+auto parse_feather(generator<chunk_ptr> input, operator_control_plane& ctrl)
+  -> generator<table_slice> {
+  auto byte_reader = make_byte_reader(std::move(input));
+  auto listener = std::make_shared<CallbackListener>();
+  auto stream_decoder = arrow::ipc::StreamDecoder(listener);
+
+  while (true) {
+    auto required_size
+      = detail::narrow_cast<size_t>(stream_decoder.next_required_size());
+    auto payload = byte_reader(required_size);
+
+    if (!payload) {
+      co_yield {};
+      continue;
+    }
+
+    const auto done = payload->size() < required_size;
+    auto decode_result
+      = stream_decoder.Consume(as_arrow_buffer(std::move(payload)));
+    if (!decode_result.ok()) {
+      diagnostic::error("failed to decode the byte stream into a record batch")
+        .note("{}", decode_result.ToString())
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+
+    while (!listener->record_batch_buffer.empty()) {
+      auto batch = listener->record_batch_buffer.front();
+      listener->record_batch_buffer.pop();
+      auto validate_status = batch->Validate();
+      TENZIR_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+      co_yield table_slice(batch);
+    }
+
+    if (done) {
+      TENZIR_INFO("returning");
+      co_return;
+    }
+  }
+}
+
+auto print_feather(
+  table_slice input, operator_control_plane& ctrl,
+  const std::shared_ptr<arrow::ipc::RecordBatchWriter>& stream_writer,
+  const std::shared_ptr<arrow::io::BufferOutputStream>& sink)
+  -> generator<chunk_ptr> {
+  auto batch = to_record_batch(input);
+  auto validate_status = batch->Validate();
+  TENZIR_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+
+  auto stream_writer_status = stream_writer->WriteRecordBatch(*batch);
+  if (!stream_writer_status.ok()) {
+    diagnostic::error("failed to write a record batch to the stream")
+      .note("{}", stream_writer_status.ToString())
+      .emit(ctrl.diagnostics());
+  }
+  auto finished_buffer_result = sink->Finish();
+  if (!finished_buffer_result.ok()) {
+    diagnostic::error("failed to close the stream and return the buffer")
+      .note("{}", finished_buffer_result.status().ToString())
+      .emit(ctrl.diagnostics());
+    co_return;
+  }
+  auto finished_buffer = finished_buffer_result.ValueOrDie();
+  ;
+  auto chunk = chunk::make(finished_buffer);
+  co_yield chunk;
+  auto reset_buffer_result
+    = sink->Reset(); // the buffer is reinit with newly allocated memory because
+                     // the API does not offer a Reset that just clears the
+                     // original data
```

Comment:
Please put the comment in a full sentence before the code that affects it; putting it after the code in the same line usually breaks code formatting with clang-format.

Thanks for the explanatory comment here, I think that really helps! 🙏